techlibs/lattice/synth_lattice.cc

```diff
@@ -0,0 +1,574 @@
+				if (nowidelut)
+					abc9_opts += " -maxlut 4";
+				if (dff)
+					abc9_opts += " -dff";
+				run("abc9" + abc9_opts);
+			} else {
+				std::string abc_args = " -dress";
+				if (nowidelut)
+					abc_args += " -lut 4";
+				else
+					abc_args += " -lut 4:7";
+				if (dff)
+					abc_args += " -dff";
+				run("abc" + abc_args);
+			}
+			run("clean");
+		}
+
+		if (check_label("map_cells"))
+		{
+			if (help_mode)
+				run("techmap -map +/lattice/cells_map.v", "(skip if -vpr)");
+			else if (!vpr)
+				run("techmap -map +/lattice/cells_map.v");
+			run("opt_lut_ins -tech lattice");
+			run("clean");
+		}
+
+		if (check_label("check"))
+		{
+			run("autoname");
+			run("hierarchy -check");
+			run("stat");
+			run("check -noinit");
+			run("blackbox =A:whitebox");
+		}
+
+		if (check_label("blif"))
+		{
+			if (!blif_file.empty() || help_mode) {
+				if (vpr || help_mode) {
+					run(stringf("opt_clean -purge"),
+							"                                 (vpr mode)");
+					run(stringf("write_blif -attr -cname -conn -param %s",
+							help_mode ? "<file-name>" : blif_file.c_str()),
+							" (vpr mode)");
+				}
+				if (!vpr)
+					run(stringf("write_blif -gates -attr -param %s",
+							help_mode ? "<file-name>" : blif_file.c_str()),
+							"       (non-vpr mode)");
+			}
+		}
+
+		if (check_label("edif"))
+		{
+			if (!edif_file.empty() || help_mode)
+				run(stringf("write_edif %s", help_mode ? "<file-name>" : edif_file.c_str()));
+		}
+
+		if (check_label("json"))
+		{
+			if (!json_file.empty() || help_mode)
+				run(stringf("write_json %s", help_mode ? "<file-name>" : json_file.c_str()));
+		}
+	}
+} SynthLatticePass;
+
+/*
+struct SynthEcp5Pass : public Pass
+{
+	SynthEcp5Pass() : Pass("synth_ecp5", "synthesis for ECP5 FPGAs") { }
+
+	void execute(std::vector<std::string> args, RTLIL::Design *design) override
+	{
+		args[0] = "synth_lattice";
+		args.insert(args.begin()+1, std::string());
+		args.insert(args.begin()+1, std::string());
+		args[1] = "-family";
+		args[2] = "ecp5";
+		Pass::call(design, args);
+	}
+} SynthEcp5Pass;
+*/
+
+struct SynthMachXO2Pass : public Pass
+{
+	SynthMachXO2Pass() : Pass("synth_machxo2", "synthesis for MachXO2 FPGAs.") { }
+
+	void execute(std::vector<std::string> args, RTLIL::Design *design) override
+	{
+		args[0] = "synth_lattice";
+		args.insert(args.begin()+1, std::string());
+		args.insert(args.begin()+1, std::string());
+		args[1] = "-family";
+		args[2] = "xo2";
+		Pass::call(design, args);
+	}
+} SynthMachXO2Pass;
+
+struct SynthMachXO3Pass : public Pass
+{
+	SynthMachXO3Pass() : Pass("synth_machxo3", "synthesis for MachXO3 FPGAs.") { }
+
+	void execute(std::vector<std::string> args, RTLIL::Design *design) override
+	{
+		args[0] = "synth_lattice";
+		args.insert(args.begin()+1, std::string());
+		args.insert(args.begin()+1, std::string());
+		args[1] = "-family";
+		args[2] = "xo3";
+		Pass::call(design, args);
+	}
+} SynthMachXO3Pass;
+
+struct SynthMachXO3DPass : public Pass
```

Comment:
Do we need this `synth_machxo3d` shim? There was never a `synth_machxo3d` pass to be backwards compatible with in the first place...

Reply:
synth_machxo3 and synth_machxo3d are more here for user convenience then for backward compatibility

Reply:
I don't think they're needed, we don't have `synth_xc7` and `synth_ultrascale` for example

Reply:
in that case I would remove synth_machxo2 as well since parameters are anyway not fully compatible and it would require user to change old scripts anyway

Reply:
yes, given the relatively low number of users of the current machxo2 flow that sounds reasonable